### PR TITLE
trees: Support soft voting using class proportions

### LINKED ISSUE #12: RFC: Inconsistent results when using `max_depth` - support soft voting
As mentioned in #9 , when limiting the depth of a RFC using `max_depth`, there are inconsistencies in the predictions. I've discussed the same issue here: https://github.com/nok/sklearn-porter/issues/52 

I've dived a bit into the issue and figured that it is indeed what I expected: The final leave probabilities are only 0 and 1 if the tree if fully grown. Therefore I thought I make a separate issue here to leave the information somewhere.

https://stackoverflow.com/questions/30814231/using-the-predict-proba-function-of-randomforestclassifier-in-the-safe-and-rig
 (see comments)

> 1) About prediction precision: I insist but this is not a question of number of trees. Even with a single decision tree you should be able to get probability predicitions with more than one digits. A decision tree aims at clustering the inputs based on some rules (the decision), and these clusters are the leafs of the tree. If you have a leaf with 2 non-spam emails and one spam email from your training data, then the probability prediction for any email that belongs to this leaf/cluster (with regards to the rules established by fitting the model), is : 1/3 for spam and 2/3 for non-spam. – Sebastien Jun 20 '15 at 14:49
>2) About the dependencies in predictions: Again Sklearn definition gives the answer : the probability is computed with regards to the leaf (corresponding to your email to test) 's characteristics : the number of instances of each class in this leaf. This is set when your model is fitted, so it only depends on the training data. In conclusion : the result is the probability of instance 1 to spam with 60% whatever the other 9 instances' probabilities are. – Sebastien Jun 20 '15 at 15:00

and 

https://stats.stackexchange.com/questions/193424/is-decision-tree-output-a-prediction-or-class-probabilities

>So depending on implementation: predicted probability is either (a) the mean terminal leaf probability across all trees or (b) the fraction of trees voting either class. If out-of-bag(OOB) prediction, then only in trees where sample is OOB. For a single fully grown tree, I would guess the predicted probability only could be 0 or 1 for any class, because all terminal nodes are pure(same label). If the single tree is not fully grown and/or more trees are grown, then the predicted probability can be a positive rational number from 0 to 1.

As I assume that it is quite a bit of re-work to integrate this changes without losing too much performance (e.g. returning pointers to arrays instead of class indexes by the individual trees), I don't want to force you to fix it just for me or so ;-) I just wanted to mention it here. So it's fine with me if it remains like it is for now.

### PR DESCRIPTION
**Pull request by jonnor**
Add (optional) support for soft voting. Fixes #12 - finally :)

Sometimes the hard majority gives unacceptable loss in performance. For those cases, it is now possible to switch to soft voting instead. This also gives smoother probabilities. Of course this comes at the cost of model size, as we must store the class proportion in the leaves, not just a class index. Inference time also goes up a bit, but I expect that impact to be less noticable in most cases.

Enabled by passing leaf_bits=N to convert(), where N is a value between 1-8. This is used to quantize the leaf values. Lower leaf_bits means fewer unique values in the leaves, which increases leaf de-duplication, and reduces the size of the model. How effective this is depends a lot on the number of classes. For fewer classes, it works better.

TODO:

- [x] Also support for inline inference mode
- [x] Add an example showing how leaf_bits parameter influences model size and predictive accuracy
- [x] Make sure documentation covers the new parameter

Support for other leaf de-duplication/compression strategies will come later.

### CODE REVIEW

**Comment by jonnor** [comment]
The results on model size here are very promising. It seems like leaf deduplication works pretty well with 4-8 bit leaves.
